### ISSUE TITLE
Linux .Net Core 3.0 "Access to the path 'blah/obj/Debug/netcoreapp3.0/blah' is denied."

### ISSUE THREAD

**Issue by pmunts** [dotnet/core-setup#8511]
On Linux, I install the .Net Core 3.0 SDK to /usr/local/lib/dotnet-sdk-3.0.100 and symlink /usr/local/lib/dotnet to it:

```
lrwxrwxrwx 1 root root     18 Sep 27 17:32 dotnet -> dotnet-sdk-3.0.100
dr-xr-xr-x 7 root root   4096 Sep 27 17:48 dotnet-sdk-3.0.100
```
If I write protect the installation directory with:

`sudo chmod -R ugo-w /usr/local/lib/dotnet-sdk-3.0.100`

When I build a project with `dotnet publish` I get long stream of error messages beginning with:

```
/usr/local/lib/dotnet-sdk-3.0.100/sdk/3.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(370,5): error MSB4018: The "CreateAppHost" task failed unexpectedly. [/home/pmunts/zz/zz.csproj]
/usr/local/lib/dotnet-sdk-3.0.100/sdk/3.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(370,5): error MSB4018: System.UnauthorizedAccessException: Access to the path '/home/pmunts/zz/obj/Debug/netcoreapp3.0/zz' is denied. [/home/pmunts/zz/zz.csproj]
```

But if I relax the permissions of the installation directory with:

`sudo chmod -R u+w /usr/local/lib/dotnet-sdk-3.0.100`

the project will build correctly.

**Comment by vitek-karas** [comment]
The original change to use `File.Copy` (which will copy permissions as well) was done here: https://github.com/dotnet/sdk/pull/1338. As @dasMulli pointed out this was for `apphost` copied from the NuGet cache folder. Currently the `apphost` can actually come from two different places:
* If the build is targeting the exact same platform as the build is running on, the `apphost` will typically come from installed SDK folder (this is because SDK installer includes the `apphost` for the "current" RID).
* If the build is targeting any other RID (for example building for linux-arm on linux-x64), the `apphost` will still come from NuGet cache - the SDK in this case restores (downloads) the necessary runtime packs for the target RID.

So in a way we need both "Fixes" - we need the `File.Copy` for the NuGet case, but we also need to make it writable for the SDK install location case.

I haven't played with the possible fix, but I think it might be as simple as adding the write permission for the current user right after the `File.Copy`. This should succeed since the file is in the intermediate folder (`obj`) which the current user should have write access to (otherwise build will fail much earlier anyway).

/cc @swaroop-sridhar

**Comment by dagood** [comment]
> So in a way we need both "Fixes" - we need the `File.Copy` for the NuGet case, but we also need to make it writable for the SDK install location case.

I don't understand why--if we think we know what permissions the dev probably wants, doesn't a simple `chmod` in both cases work? Wouldn't it be more confusing for a dev to get different permissions based on what platform they're building on (whether it comes from NuGet cache or local install)?

It seems to me that https://github.com/dotnet/core-setup/pull/8510 already fixes this.

**Comment by vitek-karas** [comment]
The problem in this issue is that the SDK fail to write to the file - because it doesn't have write access. I do agree that we don't want different permissions based on a scenario. Looking at this some more, if we simply moved the change in #8510 right after the file copy, it should fix both cases.

**Comment by am11** [comment]
I had a similar issue with a build container, where we start build containers for different language runtimes with uid of host (so inode of produced binary belong to host uid and can be used elsewhere in host, or runtime containers started by the host with same uid).

The user inside the build container is unknown and $HOME-less, such that whoami, sudo etc. do not function, and not needed in the minimal build container either.

Out of 11 language runtimes (incl. java, lua, c++, nim, c..), dotnet-core was the only unhappy camper due to this issue and heavy reliance on `$HOME` without fallbacks. Manually setting the HOME to `/tmp` fixed the bundle extraction (#7940; in certain scenarios we do not care about the `/tmp` retention policy).

**Comment by swaroop-sridhar** [comment]
Keeping the issue open until the change is in 3.1

**Comment by swaroop-sridhar** [comment]
Fixed by 10d8d09

**Comment by am11** [comment]
This is now available in latest 3.1 nightly build:

```sh
docker build -t dotnet-3.1-host-perms - << EOF
  FROM mcr.microsoft.com/dotnet/core-nightly/sdk:3.1-alpine
  RUN dotnet new console -n app > /dev/null 2>&1 &&\
      dotnet build app -o out   > /dev/null 2>&1 &&\
      stat -c %a out/app
EOF

# => 755
# vs.

docker build -t dotnet-3.0-host-perms - << EOF
  FROM mcr.microsoft.com/dotnet/core/sdk:3.0-alpine
  RUN dotnet new console -n app > /dev/null 2>&1 &&\
      dotnet build app -o out   > /dev/null 2>&1 &&\
      stat -c %a out/app
EOF

# => 744
```
